workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/WorkflowRendering.kt

```diff
@@ -0,0 +1,201 @@
+
+      // If we're being skipped it means the WorkflowRendering is either going away itself or about
+      // to switch to an incompatible rendering – either way, this lifecycle is dead.
+      // Note that we can't transition from INITIALIZED to DESTROYED – the LifecycelRegistry will
+      // throw. WorkflowLifecycleOwner has this same check.
+      if (lifecycleOwner.registry.currentState != INITIALIZED) {
+        lifecycleOwner.registry.currentState = DESTROYED
+      }
+    }
+  }
+
+  return lifecycleOwner
+}
+
+/**
+ * This is effectively the logic of [WorkflowViewStub], but translated into Compose idioms.
+ * This approach has a few advantages:
+ *
+ *  - Avoids extra custom views required to host `WorkflowViewStub` inside a Composition. Its trick
+ *    of replacing itself in its parent doesn't play nicely with Compose.
+ *  - Allows us to pass the correct parent view for inflation (the root of the composition).
+ *  - Avoids `WorkflowViewStub` having to do its own lookup to find the correct [ViewFactory], since
+ *    we already have the correct one.
+ *  - Propagate the current [LifecycleOwner] from [LocalLifecycleOwner] by setting it as the
+ *    [ViewTreeLifecycleOwner] on the view.
+ *
+ * Like `WorkflowViewStub`, this function uses the [viewFactory] to create and memoize a [View] to
+ * display the [rendering], keeps it updated with the latest [rendering] and [viewEnvironment], and
+ * adds it to the composition.
+ */
+@OptIn(WorkflowUiExperimentalApi::class)
+@Composable private fun <R : Any> ViewFactoryAndroidView(
+  viewFactory: ViewFactory<R>,
+  rendering: R,
+  viewEnvironment: ViewEnvironment,
+  modifier: Modifier
+) {
+  val lifecycleOwner = LocalLifecycleOwner.current
+
+  AndroidView(
+    modifier = modifier,
+    factory = { context ->
+      // We pass in a null container because the container isn't a View, it's a composable. The
+      // compose machinery will generate an intermediate view that it ends up adding this to but
+      // we don't have access to that.
+      viewFactory.buildView(rendering, viewEnvironment, context, container = null)
+        .also { view ->
+          // Mirrors the check done in ViewRegistry.buildView.
+          checkNotNull(view.getShowRendering<Any>()) {
+            "View.bindShowRendering should have been called for $view, typically by the " +
+              "${ViewFactory::class.java.name} that created it."
+          }
+
+          // Unfortunately AndroidView doesn't propagate this itself.
+          ViewTreeLifecycleOwner.set(view, lifecycleOwner)
+          // We don't propagate the (non-compose) SavedStateRegistryOwner, or the (compose)
+          // SaveableStateRegistry, because currently all our navigation is implemented as Android
+          // views, which only sets the state registry owner on a view that is a parent of this
+          // composition.
+          // If we want to support compose-based navigation using SaveableStateRegistry, we'd need
```

Comment:
Should we track this?

Reply:
I don't think so. If we want to re-implement the containers in compose, that'll be a huge project and this will be one tiny little detail. Should I take the comment out?

Reply:
Yeah.